src/main/scala/sparkz/core/network/peer/PeerManager.scala

```diff
@@ -40,22 +40,22 @@ class PeerManager(settings: SparkzSettings, sparkzContext: SparkzContext) extend
 
     case ConfirmConnection(connectionId, handlerRef) =>
       log.info(s"Connection confirmation request: $connectionId")
-      if (peerDatabase.isBlacklisted(connectionId.remoteAddress)) sender() ! ConnectionDenied(connectionId, handlerRef)
+      if (peerDatabase.isBlacklisted(connectionId.remoteAddress.getAddress)) sender() ! ConnectionDenied(connectionId, handlerRef)
       else sender() ! ConnectionConfirmed(connectionId, handlerRef)
 
-    case AddOrUpdatePeer(peerInfo) =>
+    case AddOrUpdatePeer(peerInfo, source) =>
       // We have connected to a peer and got his peerInfo from him
-      if (!isSelf(peerInfo.peerSpec)) peerDatabase.addOrUpdateKnownPeer(peerInfo)
-
-    case Penalize(peer, penaltyType) =>
-      log.info(s"$peer penalized, penalty: $penaltyType")
-      if (peerDatabase.peerPenaltyScoreOverThreshold(peer, penaltyType)) {
-        log.info(s"$peer blacklisted")
-        peerDatabase.addToBlacklist(peer, penaltyType)
-        sender() ! Blacklisted(peer)
+      if (!isSelf(peerInfo.peerSpec)) peerDatabase.addOrUpdateKnownPeer(peerInfo, source)
+
+    case Penalize(peerAddress, penaltyType) =>
+      log.info(s"$peerAddress penalized, penalty: $penaltyType")
+      if (peerDatabase.peerPenaltyScoreOverThreshold(peerAddress, penaltyType) && !isKnownPeer(peerAddress)) {
```

Comment:
why do we ignore known peers?
In case it started do malicious actions -> penalize
In case user decided to add it to blacklisted -> add to blacklisted.

In general: check that known peer properly works we penalization logic